sparse/impl/KokkosSparse_gmres_numeric_impl.hpp

```diff
@@ -0,0 +1,391 @@
+                                  Solver terminated without convergence.");
+        }
+        if (karith::isNan(ST(shortRelRes))) {
+          throw std::runtime_error(
+              "gmres: Relative residual is nan. Terminating solver.");
+        }
+
+        // If short residual converged, or time to restart, check true residual
+        if (shortRelRes < tol || j == m - 1) {
+          // Compute least squares soln with Givens rotation:
+          auto GLsSolnSub_h = Kokkos::subview(
+              GLsSoln_h, Kokkos::ALL,
+              0);  // Original view has rank 2, need a rank 1 here.
+          auto GVecSub_h = Kokkos::subview(GVec_h, Kokkos::make_pair(0, m));
+          Kokkos::deep_copy(GLsSolnSub_h,
+                            GVecSub_h);  // Copy LS rhs vec for triangle solve.
+          auto GLsSolnSub2_h = Kokkos::subview(
+              GLsSoln_h, Kokkos::make_pair(0, j + 1), Kokkos::ALL);
+          auto H_Sub_h = Kokkos::subview(H_h, Kokkos::make_pair(0, j + 1),
+                                         Kokkos::make_pair(0, j + 1));
+          {
+            // Hack to get around uninstantiated trsm for layoutleft
+            KokkosBlas::Impl::SerialTrsm_Invoke("L", "U", "N", "N", one,
+                                                H_Sub_h, GLsSolnSub2_h);
+          }
+          Kokkos::deep_copy(GLsSoln, GLsSoln_h);
+
+          // Update solution and compute residual with Givens:
+          VSub = Kokkos::subview(V, Kokkos::ALL, Kokkos::make_pair(0, j + 1));
+          Kokkos::deep_copy(
+              Xiter,
+              X);  // Can't overwrite X with intermediate solution.
+          auto GLsSolnSub3 =
+              Kokkos::subview(GLsSoln, Kokkos::make_pair(0, j + 1), 0);
+          if (precond) {  // Apply right prec to correct soln.
+            KokkosBlas::gemv("N", one, VSub, GLsSolnSub3, zero,
+                             Wj);                      // wj = V(1:j+1)*lsSoln
+            precond->apply(Wj, Xiter, "N", one, one);  // Xiter = M*wj + X
+          } else {
+            KokkosBlas::gemv("N", one, VSub, GLsSolnSub3, one,
+                             Xiter);  // x_iter = x + V(1:j+1)*lsSoln
+          }
+          KokkosSparse::spmv("N", one, A, Xiter, zero, Wj);  // wj = Ax
+          Kokkos::deep_copy(Res, B);                         // Reset r=b.
+          KokkosBlas::axpy(-one, Wj, Res);                   // r = b-Ax.
+          trueRes = KokkosBlas::nrm2(Res);
+          relRes  = trueRes / nrmB;
+          if (verbose) {
+            std::cout << "True relative residual for iteration "
+                      << j + (cycle * m) << " is : " << relRes << std::endl;
+          }
+          numIters = j + 1;
+
+          if (relRes < tol) {
+            converged = true;
+            Kokkos::deep_copy(
+                X, Xiter);  // Final solution is the iteration solution.
+            break;          // End Arnoldi iteration.
+          } else if (shortRelRes < 1e-30) {
```

Comment:
Probably this tolerance should be adjusted like the others I pointed out.  (I don't know how I came up with 1e-30??)